src/components/TranslateBoxes.js

```diff
@@ -3,27 +3,62 @@ import Grid from '@material-ui/core/Grid';
 import Paper from '@material-ui/core/Paper';
 import TextField from '@material-ui/core/TextField';
 import Box from '@material-ui/core/Box';
+import { Typography, Tooltip, Link } from '@material-ui/core';
 
 
 export default function TranslateBoxes({input, handleInputChange, output, errors, classes}){
-    function addErrorsToOutput(output, errors){
-        if (errors && Object.entries(errors).length !== 0){
-            let finalOutput = output
 
-            // go through output text and find any strings with pattern $$E_$$
-            const regexp = /\$\$E.\$\$/g
-            const matches = [...output.matchAll(regexp)];
+    /**
+     * Function takes output and errors and returns a list of typography and tooltip
+     * elements to be displayed in the output area. Errors are found in the output and replaced
+     * with "<err>" as well as a tooltip explaining the error.
+     * @param {response translation from backend} output 
+     * @param {response errors from backend} errors 
+     */
+    function createOutputTypography(output, errors){
+        const regexp = /\$\$E.\$\$/g
+        const matches = [...output.matchAll(regexp)];
 
+        if (matches.length > 0 && errors && Object.entries(errors).length !== 0){
+            console.log(matches)
+            let prev = 0;
+            let finalOutput = []
             for (const match of matches){
-                const errorKey = match[0].substring(2).slice(0,-2)
-                const errorMessage = errors[errorKey]["errorMessage"]
-                finalOutput = finalOutput.replace(match, errorMessage)
+                // Add text before error
+                let beforeText = output.substring(prev, match["index"])
+                finalOutput.push(
+                    <Typography className={classes.outText} key={prev}>
+                        {beforeText}
+                    </Typography>
+                )
+
+                // Add error text
+                let errorKey = match[0].substring(2).slice(0,-2)
```

Comment:
This looks like something that could be done in a single `substring` or `slice` call. Alternatively, you could add the `@@` symbols either on the frontend or the backend in the errors object keys to make the match be direct 1:1.

Reply:
Could also use capture groups in the regexp!

Reply:
Both substring and slice are necessary in order to get at it from the front and back (in order to accomodate multiple digits. I was not able to figure out how to use capture groups in javascript but I can spend more time later.

Reply:
`"$$E03$$".slice(2, -2);` Seems to correctly return `E03` for me.